File: packages/cli/src/meeting-control.ts
import { writeFileSync } from "node:fs";
import { join } from "node:path";

// One-shot control signal for the shyn-meeting agent: the agent polls
// meeting-control.json each tick and consumes it (deletes after read).
// "start" begins a recording the detector would never start on its own;
// "stop" ends the session and transcribes; "cancel" ends and discards.
function write(home: string, action: "start" | "stop" | "cancel",
               title?: string, attendees?: string[]): void {
  writeFileSync(join(home, "meeting-control.json"),
    JSON.stringify({ action, title, attendees: attendees?.length ? attendees : undefined,
                     ts: Math.floor(Date.now() / 1000) }) + "\n");
}
// Detection is audio-shaped: the commit gate requires far-side voice on the
// system channel, so a room — where every voice arrives on the microphone —
// is invisible to it by construction. This is the way in. The optional title
// is the only name such a recording can get; there is no tab and no calendar
// entry to infer one from.
export const requestMeetingStart = (home: string, title?: string, attendees: string[] = []) =>
  write(home, "start", title, attendees);

// `shyn meeting start <title words…> [--with "a, b"]`. Everything that is
// not the --with flag and its value is the title. The roster is the one
// fact about a room recording shyn cannot infer, so it can be typed here
// the same way the menu bar form asks for it.
export function parseMeetingStartArgs(rest: string[]): { title: string | undefined; attendees: string[] } {
  const words: string[] = [];
  let attendees: string[] = [];
  for (let i = 0; i < rest.length; i++) {
    if (rest[i] === "--with") {
      attendees = (rest[i + 1] ?? "").split(",").map((s) => s.trim()).filter(Boolean);
      i++;
      continue;
    }
    words.push(rest[i]);
  }
  const title = words.join(" ").trim() || undefined;
  return { title, attendees };
}
export const requestMeetingStop = (home: string) => write(home, "stop");
export const requestMeetingCancel = (home: string) => write(home, "cancel");

File: packages/capture-agent/Sources/CaptureCore/MeetingControl.swift
import Foundation

// One-shot control signal for the meeting agent, written by `shyn meeting
// <start|stop|cancel>` and by the menu bar, consumed (deleted) by the agent on
// its next tick.
//
// `start` exists because detection is audio-shaped and always will be: the
// commit gate requires far-side voice on the SYSTEM channel, deliberately, so
// an idle browser tab or a calendar entry can never commit silence. That rule
// is right for calls and useless for a room — an in-person meeting puts every
// voice on the microphone and nothing on the system channel, so no amount of
// tuning would ever make the detector notice it. An explicit start IS the
// verification, and it is the only way to record a room, a speakerphone, or
// anything else shyn cannot hear itself into.
//
// Parsing lives in CaptureCore rather than next to the file IO because that is
// where the test target can reach it — stop/cancel had been parsed in
// shyn-meeting since Task 10 and never had a single test.

public enum MeetingAction: String, Sendable { case start, stop, cancel }

public struct MeetingControl: Equatable, Sendable {
    public let action: MeetingAction
    /// Only ever set for `.start`: what to call the recording. The detector
    /// has no tab or calendar entry to name a manual session after.
    public let title: String?
    /// Only ever set for `.start`: who was in the room, as typed into the menu
    /// bar form or `shyn meeting start --with`. shyn cannot infer this for a
    /// room (every voice is on one channel), and it is the one thing that
    /// makes such a recording findable later. Trimmed, deduplicated, capped.
    public let attendees: [String]
    public init(action: MeetingAction, title: String?, attendees: [String] = []) {
        self.action = action
        self.title = title
        self.attendees = attendees
    }
}

/// A name is a name, not a sentence; and a roster of 30 is already a lecture.
let maxManualAttendeeLength = 60
let maxManualAttendees = 30

func cleanAttendees(_ raw: Any?) -> [String] {
    guard let list = raw as? [Any] else { return [] }
    var seen = Set<String>(), out: [String] = []
    for item in list {
        guard let s = item as? String else { continue }
        let t = String(s.trimmingCharacters(in: .whitespacesAndNewlines).prefix(maxManualAttendeeLength))
        guard !t.isEmpty, seen.insert(t.lowercased()).inserted else { continue }
        out.append(t)
        if out.count == maxManualAttendees { break }
    }
    return out
}

/// Longest title we will put on a document. Matches cleanMeetingWindowTitle's
/// ceiling for the same reason: a title is not a place for a paragraph.
let maxManualTitleLength = 120

/// Parses a control file's bytes. Returns nil for anything unrecognised —
/// garbage, an empty file, or a verb from a newer CLI than this agent. The
/// caller deletes the file regardless, so an unparseable control is skipped
/// rather than retried forever.
public func parseMeetingControl(_ data: Data) -> MeetingControl? {
    guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let raw = obj["action"] as? String,
          let action = MeetingAction(rawValue: raw) else { return nil }
    guard action == .start else { return MeetingControl(action: action, title: nil) }
    let trimmed = (obj["title"] as? String)?
        .trimmingCharacters(in: .whitespacesAndNewlines)
    let attendees = cleanAttendees(obj["attendees"])
    guard let t = trimmed, !t.isEmpty else {
        return MeetingControl(action: .start, title: nil, attendees: attendees)
    }
    return MeetingControl(action: .start,
                          title: t.count <= maxManualTitleLength
                              ? t : String(t.prefix(maxManualTitleLength)),
                          attendees: attendees)
}
